File: data_structures/linked_lists/linked_list.js
const Link = require('./link.js');

class LinkedList {
  constructor(val) {
    this.head = new Link(val);
  }

  add(val) {
    let current = this.head;
    let prev;
    while (current.next !== null) {
      prev = current;
      current = current.next;
    }
    prev = current;
    current.next = new Link(val, null, prev);
  }

  returnList() {
    let current = this.head;
    let vals = [];
    while (current.next !== null) {
      vals.push(current.val);
      current = current.next;
    }
    vals.push(current.val);
    return vals;
  }
}

module.exports = LinkedList;

File: data_structures/linked_lists/linked_list.rb
require_relative 'link'

class LinkedList
  attr_accessor :head, :tail

  def initialize(val)
    @head = Link.new(val)
  end

  def add(val)
    current = @head
    while current.next_link != nil
      prev = current
      current = current.next_link
    end
    prev = current
    current.next_link = Link.new(val, nil, prev)
  end

  def delete(val)
    current.next_link = @head
    if current.val == val
      @head = current.next_link
    else
      while (current.next_link != nil) && (current.next_link.val != val)
        current = current.next_link
      end
      unless current.next_link = nil
        current.next_link = current.next_link.next_link
      end
    end
  end

  def return_list
    current = @head
    vals = []
    until current.next_link == nil
      vals << current.val
      current = current.next_link
    end
    vals << current.val
    vals
  end
end
